### PR TITLE
[DOCS] ES|QL implicit casting

### PR DESCRIPTION
**Pull request by fang-xing-esql**
Add docs for implicit casting.

### CODE REVIEW

File: docs/reference/esql/implicit-casting.asciidoc [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+----
+from employees
+| eval dd_ns1=date_diff("day", to_datetime("2023-12-02T11:00:00.00Z"), birth_date)
+| sort emp_no
+| keep dd_ns1
+| limit 1
+;
+----
+
+Implicit casting is to improve the usability, the string literals will be converted to the target data type automatically. This is mostly useful when the target data type is datetime, ip, version or a geo spatial, as it is natural to specify them as a string in queries. The query above can be coded like below without calling the to_datetime function to do conversions manually.
```
Comment (leemthompo) [inline docs/reference/esql/implicit-casting.asciidoc:22]:
```suggestion
Implicit casting improves usability, by automatically converting string literals to the target data type. This is most useful when the target data type is `datetime`, `ip`, `version` or a geo spatial. It is natural to specify these as a string in queries.

The first query can be coded without calling the `to_datetime` function, as follows:
```

File: docs/reference/esql/implicit-casting.asciidoc [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+----
+from employees
+| eval dd_ns1=date_diff("day", "2023-12-02T11:00:00.00Z", birth_date)
+| sort emp_no
+| keep dd_ns1
+| limit 1
+;
+----
+
+In the table below, are the scenarios that implicit casting is supported.
```
Comment (leemthompo) [inline docs/reference/esql/implicit-casting.asciidoc:34]:
```suggestion
[float]
[[esql-implicit-casting]]
=== Implicit casting support

The following table details which {esql} operations support implicit casting for different data types.
```

**Review by leemthompo — COMMENTED** [review]
This is looking good @fang-xing-esql, just a few suggestions for clarity and presentation so far :)

We would definitely want to link this page from the [type conversion functions section](https://www.elastic.co/guide/en/elasticsearch/reference/current/esql-functions-operators.html#esql-type-conversion-functions), perhaps in a `TIP` callout.

**Comment by fang-xing-esql** [comment]
> This is looking good @fang-xing-esql, just a few suggestions for clarity and presentation so far :)
> 
> We would definitely want to link this page from the [type conversion functions section](https://www.elastic.co/guide/en/elasticsearch/reference/current/esql-functions-operators.html#esql-type-conversion-functions), perhaps in a `TIP` callout.

Thank you for reviewing Liam! The last commit is updated according to the comments, could you please take another look?

File: docs/reference/esql/implicit-casting.asciidoc [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+The first query can be coded without calling the `to_datetime` function, as follows:
+
+[source.merge.styled,esql]
+----
+from employees
+| eval dd_ns1=date_diff("day", "2023-12-02T11:00:00.00Z", birth_date)
+| sort emp_no
+| keep dd_ns1
+| limit 1
+;
```
Comment (leemthompo) [inline docs/reference/esql/implicit-casting.asciidoc:33]:
```suggestion
```

**Review by leemthompo — COMMENTED** [review]
Couple minor things, plus let's add link in `esql-language.asciidoc`, you already have the `include` statement, but just need link text so it appears in the list on that page :)

Once addressed should be good to go, thank you!

File: docs/reference/esql/esql-language.asciidoc [hunk 4]
```diff
@@ -23,3 +23,4 @@ include::esql-index-options.asciidoc[]
 include::multivalued-fields.asciidoc[]
 include::esql-process-data-with-dissect-grok.asciidoc[]
 include::esql-enrich-data.asciidoc[]
+include::implicit-casting.asciidoc[]
```
Comment (leemthompo) [inline docs/reference/esql/esql-language.asciidoc:26]: will need to add link <<esql-implicit-casting>> above along with others :)

**Comment by fang-xing-esql** [comment]
> Couple minor things, plus let's add link in `esql-language.asciidoc`, you already have the `include` statement, but just need link text so it appears in the list on that page :)
> 
> Once addressed should be good to go, thank you!

Thank you for double checking again @leemthompo ! All of the comments are addressed.

**Review by leemthompo — APPROVED** [review]
LGTM from a docs perspective. Thanks for iterating!

**Pull request merged**